File: 2-add-two-numbers/2-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    /**
     * IDEA(s):
     * Naive Solution:
     * - Loop over each linked list, convert the numbers into strings, parse the
     *   strings as integers, add theh integers, then convert the integers into
     *   a linked list to be returned.
     * - WON'T WORK: Max int (in Java) is ~2 billion. That's 9 digits. We'll get
     *   an overflow b/c this solution must accept values with up to 100 digits.
     * Beter Solution:
     * - Loop over both linked lists simultaneously. On each iteration, add the
     *   two values 
     * 
     * OUTLINE:
     * - While l1 OR l2 is not null
     *     - if 
     * 
     */ 
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        ListNode emptyHead = new ListNode();
        ListNode current = emptyHead;
        
        ListNode p1 = l1;
        ListNode p2 = l2;
        
        int sum = 0;
        int carry = 0;
        
        while (p1 != null || p2 != null || carry > 0) {
            
            int val_1 = 0;
            int val_2 = 0;
            
            if (p1 != null) {
                val_1 = p1.val;
                p1 = p1.next;
            }
            if (p2 != null) {
                val_2 = p2.val;
                p2 = p2.next;
            }

            sum = val_1 + val_2;
            sum += carry;
            carry = 0;
            System.out.println("First :: sum: " + sum + ", carry: " + carry);
            if (sum >= 10) {
                carry = (sum / 10);
                sum = (sum % 10);
            }
            System.out.println("Second :: sum: " + sum + ", carry: " + carry);
            current.next = new ListNode(sum);
            current = current.next;
        }
        
        return emptyHead.next;
    }
}

File: LeetCode/0002-add-two-numbers/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        rHead = ListNode()
        
        carry = 0
        current = rHead
        while l1 and l2:
            sum = l1.val + l2.val + carry
            newNode = ListNode(int(sum % 10))
            current.next = newNode
            current = current.next
            carry = int(sum / 10)

            l1 = l1.next
            l2 = l2.next

        while l1:
            sum = l1.val + carry
            newNode = ListNode(int(sum % 10))
            current.next = newNode
            current = current.next
            carry = int(sum / 10)

            l1 = l1.next
        while l2:
            sum = l2.val + carry
            newNode = ListNode(int(sum % 10))
            current.next = newNode
            current = current.next
            carry = int(sum / 10)

            l2 = l2.next
        
        if carry != 0:
            current.next = ListNode(carry)
        
        return rHead.next

# ***** IDEAS *****
# IDEA 1:
# > Loop over the two input lists simultaneously. For each node, add the values together and extract the second (Least Significant Digit) digit and store it in the next node in the return list, then store the Most Significant Digit in a variable as the carray. Continue until the input lists are expended.
# > The trick will be handling input lists of differing length.
#   > we need to loop over the longer list and simply transfer the values to the output list.
#   > Wait, this isn't an issue. Since the numbers are reversed, the heads of the linked list will represent the Least Significant Digit. This means that we can just start adding the numbers together until one of the lists is null. At which point we simply add the remainder of the list that still has nodes to the output list.
# ***** OUTLINE *****
# Instantiate output list (dummy node head: rHead)
# Instantiate carry = 0
# While both lists are not null,
#     sum = l1.val + 12.val + carry
#     Add a new node with .val = sum mod 10 (to extract LSD)
#     carry = sum / 10 (to extract MSD)
# If l2 not None, rHead.next = l2
# If l1 not None, rHead.next = l1
# Return dummy.next